File: companyPython/Funcionario.py
from abc import ABC, abstractmethod
import re

class Funcionario(ABC):
    def __init__(self, nome, cpf, salarioBase):
        self.nome = nome
        self.cpf = cpf
        self.salarioBase = salarioBase

    @abstractmethod
    def calculaSalario():
        pass

    @staticmethod
    def verificaCPF(cpf):
        # Retirando caracteres nao numericos do CPF.
        cpf = re.sub('[^0-9]', "", cpf)

        if (len(cpf) != 11):
            return False

        # Pegando os 9 primeiros digitos do CPF.
        cpfBase = Funcionario.criaCpfBase(cpf)
        
        # Calculando o primeiro digito verificador e validando-o.
        digVerificador = Funcionario.calculaDigitoVerificador(cpfBase, 10)
        if (digVerificador != int(cpf[9])):
            return False
        
        # Caso o digito seja valido, adicionando-o ao CPF base.
        cpfBase.append(digVerificador)

        # Calculando o segundo digito verificador e validando-o.
        digVerificador = Funcionario.calculaDigitoVerificador(cpfBase, 11)
        if (digVerificador != int(cpf[10])):
            return False
        
        return True
    
    @staticmethod
    # Retorna os 9 primeiros valores de uma String de numeros.
    def criaCpfBase(cpf):
        cpfBase = []

        for i in range(9):
            cpfBase.append(int(cpf[i]))

        return cpfBase

    @staticmethod
    # Calcula os digitos verificadores de acordo com a especificacao do professor.
    def calculaDigitoVerificador(cpf, pesoInicial):
        soma = 0

        i = 0
        for j in range(pesoInicial, 1, -1):
            soma += (cpf[i] * j)
            i += 1

        digVerificador = (soma * 10) % 11
        if (digVerificador == 10 or digVerificador == 11):
            digVerificador = 0

        return digVerificador

File: companyJava/Funcionario.java
package empresa;

public abstract class Funcionario {
	String nome;
	String CPF;
	double salarioBase;
	
	public Funcionario(String nome, String cpf, double salario) {
		this.nome = nome;
		this.CPF = cpf;
		this.salarioBase = salario;
	}
	
	protected abstract double calculaSalario();
	
	public static boolean verificaCPF(String cpf) {
		// Retirando caracteres nao numericos do CPF.
		cpf = cpf.replaceAll("[^0-9]", "");
		
		if (cpf.length() != 11) return false;
		
		// Pegando os 9 primeiros digitos do CPF.
		int cpfBase[] = criaCpfBase(cpf);
		
		// Calculando o primeiro digito verificador e vendo, a partir dele, se o CPF eh valido.
		int digVerificador = calculaDigitoVerificador(cpfBase, 10);
		if (digVerificador != Character.getNumericValue(cpf.charAt(9))) return false;
		// Caso o CPF seja valido, adiciono o primeiro digito verificador ao CPF.
		cpfBase[9] = digVerificador;
		
		// Calculando o segundo digito verificador e vendo, a partir dele, se o CPF eh valido.
		digVerificador = calculaDigitoVerificador(cpfBase, 11);
		if (digVerificador != Character.getNumericValue(cpf.charAt(10))) return false;
		
		return true;
	}
	
	// Retorna os 9 primeiros valores de uma String de numeros.
	private static int[] criaCpfBase(String cpf) {
		int cpfBase[] = new int[10];

		for (int i = 0; i < 9; i++) {
			cpfBase[i] = Character.getNumericValue(cpf.charAt(i));
		}
		
		return cpfBase;
	}
	
	// Calcula os digitos verificadores de acordo com a especificacao do professor.
	private static int calculaDigitoVerificador(int cpfBase[], int pesoInicial) {
		int soma = 0;
		
		for (int i = 0; (pesoInicial - i) >= 2; i++) {
			soma += cpfBase[i] * (pesoInicial - i);
		}

		int digVerificador = (soma * 10) % 11;
		if (digVerificador == 10 || digVerificador == 11) {
			digVerificador = 0;
		}
		
		return digVerificador;
	}
}